src/queue/queue.ts

```diff
@@ -0,0 +1,57 @@
+// import * as batch from "../batch/batch";
+import { getConfig, Config } from "../config/config";
+// import * as announcement from "../contracts/announcement";
+// import * as storage from "../storage/storage";
+import { DSNPMessage } from "../messages/messages";
+import { HexString } from "../types/Strings";
+import { NotImplementedError } from "../utilities";
+
+export type QueueId = HexString;
+
+export interface QueueInterface {
+  enqueue(dsnpMessage: DSNPMessage): Promise<QueueId>;
+  dequeue(id: QueueId): Promise<DSNPMessage>;
+  getAll(): Promise<DSNPMessage[]>;
+}
+
+/**
+ * enqueue() adds an activity pub event to the queue for later publishing to the
+ * blockchain as a batch file.
+ *
+ * @param dsnpEvent The DSNP message to queue up for batching
+ * @param opts      Optional. Configuration overrides, such as from address, if any
+ * @returns         An ID for the queued message
+ */
+export const enqueue = async (event: DSNPMessage, opts?: Config): Promise<QueueId> => {
+  const config = getConfig(opts);
+  return await config.queue.enqueue(event);
+};
+
+/**
+ * dequeue() removes an activity pub event from the queue for later publishing
+ * to the blockchain.
+ *
+ * @param id   The ID of the message to remove from the queue
+ * @param opts Optional. Configuration overrides, such as from address, if any
+ * @returns    The DSNP message removed from the queue
+ */
+export const dequeue = async (id: QueueId, opts?: Config): Promise<DSNPMessage> => {
+  const config = getConfig(opts);
+  return await config.queue.dequeue(id);
+};
+
+/**
+ * commit() creates a batch file from the current activity pub messages in the
+ * queue then clears the queue. This method is not yet implemented.
+ */
+export const commit = async (_opts?: Config): Promise<void> => {
+  throw NotImplementedError;
+
+  // const config = getConfig(opts);
+  // const events = await config.queue.getAll();
+  // const batchFile = batch.createFile(events);
+  //
+  // const hash = batch.hash(batchFile);
+  // const uri = await storage.put(batchFile, opts);
+  // await announcement.batch(uri, hash);
+};
```

Comment:
Not related to this in particular, but I wonder if this shouldn't just have a `popAll` instead of a commit. I don't like the idea that this would call the batch module or that the batch module would call this one. That feels higher level than the module.

Reply:
Good idea. I'll add a note to the story for implementing this.

Reply:
Replaced `commit()` with a `dequeueBatch()` function. I chose `dequeueBatch` over `popBatch` because the function pulls items from the front of the queue rather than the end like a stack, and I chose `dequeueBatch` over `dequeueAll` since it takes a `count` parameter and doesn't necessarily dequeue the entire list.